### PR TITLE
feat(ecologits): green-cost metrics with the Ecologits observabilty plugin for litellm proxy

### PR DESCRIPTION
**Pull request by Jesshuan**
## Relevant issues

Feature sugggestion here: #31226

## Pre-Submission checklist

**Please complete all items before asking a LiteLLM maintainer to review your PR**

- [X] I have added meaningful tests
- [X] My PR passes all unit tests on [`make test-unit`](https://docs.litellm.ai/docs/extras/contributing_code)
- [X] My PR's scope is as isolated as possible; it only solves 1 specific problem
- [x] I have requested a Greptile review by commenting `@greptileai` and received a **Confidence Score of at least 4/5** before requesting a maintainer review

## Delays in PR merge?

If you're seeing a delay in your PR being merged, ping the LiteLLM Team on [Slack (#pr-review)](https://join.slack.com/t/litellmossslack/shared_invite/zt-3o7nkuyfr-p_kbNJj8taRfXGgQI1~YyA).

## CI (LiteLLM team)

> **CI status guideline:**
>
> - 50-55 passing tests: main is stable with minor issues.
> - 45-49 passing tests: acceptable but needs attention
> - <= 40 passing tests: unstable; be careful with your merges and assess the risk.

- [ ] **Branch creation CI run**  
       Link:

- [ ] **CI run for the last commit**  
       Link:

- [ ] **Merge / cherry-pick CI run**  
       Links:

## Screenshots / Proof of Fix

- ecologits callback is set within my config.yaml, side-to-sire with langfuse and prometheus client. And i add (optionnaly) zone for each model, in 'model info'

image

- The litellm proxy is used by a personal app. Here, a Windmill orchestrator that provides workflow with automation  with an 'AI Agent' step. I've choosen a mistral ai agent:

image

- The token usage first can be see in litellm by itself, on the dashboard:

image

- The Ecologits green-cost appears in Langfuse within the 'metadata' field:
image

- ... and the payload contains the initial ecologits payload with the arguments that were used for the Ecologits API calculation:

image

- On a Grafana dashboard, with the Prometheus collector (prometheus client active), we can monitor the green-cost metrics and play with the labels availables:

... here with the label 'mistral_ai' and a label choice for the electricity mix zone ('FRA')...

image

...here with the provider 'anthropic' and a choice of a specific 'model', for example:

image

There are two lines for the two bounds (if we don't choose a label' bound') min/max, because the Ecologits API provide a min and a max for its estimations with a provider like anthropic, when the measurement is difficult.

## Type

🆕 New Feature
✅ Test

## Changes

We propose to enrich LiteLLM's existing LLM-call metrics with **green-cost metrics**, via an optional EcoLogits callback plugin. On every successful call, the plugin queries the [EcoLogits API](https://api.ecologits.ai/docs) and makes the resulting environmental impacts available to **all downstream observability callbacks** (Langfuse, Datadog, Prometheus, …).

For this first "proxy mode" version, the principle is simple:

- **Per-model zone config.** When registering a model, admins can set a `model_info` argument `ecologits_electricity_mix_zone` (e.g. `FRA`, `DEU`) reflecting where the model physically runs. If omitted, the plugin falls back to the `ECOLOGITS_ELECTRICITY_MIX_ZONE` env var, and finally to EcoLogits' own default (`WOR` — world average).

- **Per-call enrichment.** On each successful call to a provider, the plugin calls the EcoLogits API (public, or an enterprise self-hosted instance via `ECOLOGITS_API_BASE`) and attaches the impacts to the call payload. It runs in the `async_logging_hook`, *before* other callbacks see the data — so any downstream integration the user enables gets the green metrics for free. The enrichment is written where each consumer looks: `litellm_params.metadata` (Langfuse) and `standard_logging_object.metadata` (Datadog, SpendLogs). For Prometheus, dedicated green-cost Counters (energy kWh, GWP kgCO2eq, …) are exposed alongside.

- **Never fatal.** Any failure (timeout, non-200, bad payload) is swallowed — at worst the call is logged without enrichment.

See the #31226 for the motivations of this work.

A PR for the documentation part of this new callback integration will follow these days.

### CODE REVIEW

**Comment by Sameerlite** [comment]
Thanks so much for this contribution — the EcoLogits integration looks really well thought out, and the test coverage is impressive!\n\nA few things to address before this can merge:\n\n1. **Greptile score is 4/5 with 3 unresolved threads** — Greptile flagged some real concerns, including the unconditional  import at module level (breaking users without it installed), a potential double-yield in , and use of a private  internal. Could you work through those? We need a 5/5 with no open threads.\n\n2. **CI is failing** — , , , and  checks are red. Could you fix those up? If any failures are pre-existing/flaky, a quick note here would help.\n\n3. **No proof of working** — could you add some evidence in the Screenshots / Proof of Fix section? A sample log showing the EcoLogits metadata being attached to a call (e.g. Langfuse metadata output, Prometheus counter output, or a simple test run result) would go a long way.\n\nAppreciate the work on this — happy to move quickly once those are addressed!

File: litellm/integrations/ecologits.py [hunk 1]
```diff
@@ -0,0 +1,339 @@
...
+    for json_key, suffix, unit, description in ECOLOGITS_IMPACTS:
+        counters[json_key] = Counter(
+            name=f"litellm_ecologits_{suffix}_total",
+            documentation=f"Cumulative {description} ({unit}) attributed to LLM calls, from EcoLogits.",
+            labelnames=ECOLOGITS_PROMETHEUS_LABELS,
+        )
+    return counters
+
+
+_ECOLOGITS_COUNTERS = _build_ecologits_counters()
```
[the remark this answers is not in the document]
Comment (Jesshuan) [reply litellm/integrations/ecologits.py]: solved in the last commits

File: litellm/integrations/ecologits.py [hunk 2]
```diff
@@ -0,0 +1,339 @@
...
+        if isinstance(raw_value, (int, float)):
+            yield "value", raw_value
+        elif isinstance(raw_value, dict):
+            for bound in ("min", "max"):
+                if bound in raw_value:
+                    yield bound, raw_value[bound]
+
+        for bound in ("min", "max"):
+            if bound in entry:
+                yield bound, entry[bound]
```
[the remark this answers is not in the document]
Comment (Jesshuan) [reply litellm/integrations/ecologits.py]: solved too

**Comment by Jesshuan** [comment]
@Sameerlite 

Great ! Happy taht you enjoy this Ecologits integration work...

I try to solve all that you've mentioned. And i have do a PR on the litellm/docs in parallel to remove two of the three red checks here...

A little question: how to say to greptile-apps to be re-trigged after my commits resolutions ?
(The "re-trigged" button don't work for me, i think i don't have the permissions and the authentication don't work)

Perhaps tagging it ? I try : @greptile-apps

**Comment by Jesshuan** [comment]
@greptile-apps can you check the code again. I should have finish to make all the corrections

**Comment by Jesshuan** [comment]
@Sameerlite 

This PR seems technically ready... The two missing failed check steps are link to the PR submitted in parallel on the litellm/docs project: https://github.com/BerriAI/litellm-docs/pull/428

But if you think it's now totally good here... can you just tag me before really merging ?
I've sent all this integration plan to the Ecologits team and they've decided to fastly deploy a gateway for their Ecologits API, because at this time, it have no rate limits and it is totally opened without any limitations. And after that, i think taht i will have to add an optionnal private key for the Ecologits API call.
